core/store/lib/search/adapter/api/elasticsearchQuery.js

```diff
@@ -0,0 +1,110 @@
+          if (isRange.length) {
+            let rangeAttribute = catalogfilter.attribute
+            // filter by product fiunal price
+            if (rangeAttribute === 'price') {
+              rangeAttribute = 'final_price'
+            }
+            // process range filters
+            filterQr = filterQr.andFilter('range', rangeAttribute, catalogfilter.value)
+          } else {
+            // process terms filters
+            let newValue = catalogfilter.value[Object.keys(catalogfilter.value)[0]]
+            if (!Array.isArray(newValue)) {
+              newValue = [newValue]
+            }
+            if (attrPostfix === '') {
+              filterQr = filterQr.andFilter('terms', getMapping(catalogfilter.attribute), newValue)
+            } else {
+              filterQr = filterQr.andFilter('terms', catalogfilter.attribute + attrPostfix, newValue)
+            }
+          }
+        }
+      })
+      return filterQr
+    }
+
+    if (hasCatalogFilters) {
+      query = query.orFilter('bool', (b) => attrFilterBuilder(b))
+        .orFilter('bool', (b) => attrFilterBuilder(b, optionsPrfeix).filter('match', 'type_id', 'configurable')) // the queries can vary based on the product type
+    }
+  }
+
+  // Add aggregations for catalog filters
+  const allFilters = searchQuery.getAvailableFilters()
+  if (allFilters.length > 0) {
+    for (let attrToFilter of allFilters) {
+      if (attrToFilter.scope === 'catalog') {
+        if (attrToFilter.field !== 'price') {
+          query = query.aggregation('terms', getMapping(attrToFilter.field))
+          query = query.aggregation('terms', attrToFilter.field + optionsPrfeix)
+        } else {
+          query = query.aggregation('terms', attrToFilter.field)
+          query.aggregation('range', 'price', {
+            ranges: [
+              { from: 0, to: 50 },
+              { from: 50, to: 100 },
+              { from: 100, to: 150 },
+              { from: 150 }
+            ]
+          })
+        }
+      }
+    }
+  }
+
+  if (queryText !== '') {
+    query = query.andQuery('bool', b => b.orQuery('match_phrase_prefix', 'name', { query: queryText, boost: getBoosts('name'), slop: 2 })
+      .orQuery('match_phrase', 'category.name', { query: queryText, boost: getBoosts('category.name') })
+      .orQuery('match_phrase', 'short_description', { query: queryText, boost: getBoosts('short_description') })
+      .orQuery('match_phrase', 'description', { query: queryText, boost: getBoosts('description') })
+      .orQuery('bool', b => b.orQuery('terms', 'sku', queryText.split('-'))
+        .orQuery('terms', 'configurable_children.sku', queryText.split('-'))
+        .orQuery('match_phrase', 'sku', { query: queryText, boost: getBoosts('sku') })
+        .orQuery('match_phrase', 'configurable_children.sku', { query: queryText, boost: getBoosts('configurable_children.sku') }))
+    )
+  }
+
+  const queryBody = query.build()
+
+  return queryBody
+}
```

Comment:
I believe we should have a more general way to create full-text queries. I mean it would be great if the SearchQuery class can support constructing the text queries like this one to not have this logic centralized as it currently is